lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,207 @@
+          #       associations: {
+          #         author: {
+          #           data: {
+          #             type: 'user',
+          #             id: 2
+          #           }
+          #         },
+          #         second_author: {
+          #           data: nil
+          #         },
+          #         comments: {
+          #           data: [{
+          #             type: 'comment',
+          #             id: 3
+          #           },{
+          #             type: 'comment',
+          #             id: 4
+          #           }]
+          #         }
+          #       }
+          #     }
+          #   }
+          #
+          #   parse(document) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   date: '2015-12-20',
+          #     #   author_id: 2,
+          #     #   second_author_id: nil
+          #     #   comment_ids: [3, 4]
+          #     # }
+          #
+          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   published_at: '2015-12-20',
+          #     #   author_id: 2,
+          #     # }
+          #
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def parse!(document, options = {})
+            fields = parse_fields_option(options[:fields])
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            fail ArgumentError, 'Invalid payload' unless payload_valid?(document)
+
+            hash = {}
+
+            primary_data = document['data']
+
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            attributes = primary_data['attributes']
+            hash.merge!(parse_attributes(attributes, fields)) if attributes
+
+            relationships = primary_data['relationships']
+            hash.merge!(parse_relationships(relationships, fields)) if relationships
+
+            hash
+          end
+          # rubocop:enable Metrics/CyclomaticComplexity
+
+          # Safe version of parse!. Returns an empty hash instead of raising ArgumentError.
+          def parse(document, options = {})
+            parse!(document, options)
+          rescue ArgumentError
+            {}
+          end
+
+          # Checks whether a payload is compliant with the JSON API spec.
+          #
+          # @api private
+          # rubocop:disable Metrics/CyclomaticComplexity
+          def payload_valid?(payload)
+            return false unless payload.is_a?(Hash)
+            primary_data = payload['data']
+            return false unless primary_data.is_a?(Hash)
+            attributes = primary_data['attributes']
+            return false unless attributes.nil? || attributes.is_a?(Hash)
+            relationships = primary_data['relationships']
+            return false unless relationships.nil? || relationships.is_a?(Hash)
```

Comment:
Thanks again for the quick response and your work on this! I'm looking forward to getting it into production.